File: mobile/TraderDesk/android/src/main/java/com/traderdesk/TradingSolverModule.kt
package com.traderdesk

import android.util.Base64
import com.facebook.react.bridge.*
import kotlinx.coroutines.*
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.DoubleBuffer

/**
 * TradingSolverModule — React Native bridge to the Zig LP solver (.so)
 *
 * The native library (libtrading_solver.so) is loaded via System.loadLibrary.
 * JNI declarations delegate to the C-ABI functions exported by solver_mobile.zig.
 *
 * Build:
 *   The .so files for each ABI are placed in:
 *   android/src/main/jniLibs/<abi>/libtrading_solver.so
 *
 *   Supported ABIs: arm64-v8a, armeabi-v7a, x86_64
 */
class TradingSolverModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName() = "TradingSolver"

    companion object {
        init {
            System.loadLibrary("trading_solver")
        }

        // ── JNI declarations ────────────────────────────────────────────────
        // These match the C signatures in solver_mobile.zig

        @JvmStatic
        external fun nativeSolve(
            modelDescriptor: ByteArray,
            modelLen: Int,
            variables: DoubleArray,
            nVars: Int,
            outStatus: IntArray,        // [0] = status
            outScalars: DoubleArray,    // [profit, tons, cost, roi]
            outNRoutes: IntArray,       // [0] = n_routes
            outRouteTons: DoubleArray,  // [16]
            outRouteProfits: DoubleArray,
            outMargins: DoubleArray,
            outNConstraints: IntArray,  // [0] = n_constraints
            outShadowPrices: DoubleArray // [32]
        ): Int

        @JvmStatic
        external fun nativeMonteCarlo(
            modelDescriptor: ByteArray,
            modelLen: Int,
            center: DoubleArray,
            nVars: Int,
            nScenarios: Int,
            outStatus: IntArray,
            outCounts: IntArray,        // [n_scenarios, n_feasible, n_infeasible]
            outStats: DoubleArray,      // [mean, stddev, p5, p25, p50, p75, p95, min, max]
            outNSens: IntArray,         // [0] = n_sensitivity
            outSensIdx: IntArray,       // [64]
            outSensCorr: DoubleArray    // [64]
        ): Int

        @JvmStatic
        external fun nativeGetVersion(): String
    }

    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    // ── solve ──────────────────────────────────────────────────────────────────

    @ReactMethod
    fun solve(
        descriptorBase64: String,
        variablesArray: ReadableArray,
        promise: Promise
    ) {
        scope.launch {
            try {
                val descriptor = Base64.decode(descriptorBase64, Base64.DEFAULT)
                val variables = variablesArray.toDoubleArray()

                val outStatus = IntArray(1)
                val outScalars = DoubleArray(4)
                val outNRoutes = IntArray(1)
                val outRouteTons = DoubleArray(16)
                val outRouteProfits = DoubleArray(16)
                val outMargins = DoubleArray(16)
                val outNConstraints = IntArray(1)
                val outShadowPrices = DoubleArray(32)

                nativeSolve(
                    descriptor, descriptor.size,
                    variables, variables.size,
                    outStatus, outScalars,
                    outNRoutes, outRouteTons, outRouteProfits, outMargins,
                    outNConstraints, outShadowPrices
                )

                val nRoutes = outNRoutes[0]
                val nConstraints = outNConstraints[0]

                val result = Arguments.createMap().apply {
                    putInt("status", outStatus[0])
                    putDouble("profit", outScalars[0])
                    putDouble("tons", outScalars[1])
                    putDouble("cost", outScalars[2])
                    putDouble("roi", outScalars[3])
                    putInt("nRoutes", nRoutes)
                    putArray("routeTons", outRouteTons.take(nRoutes).toReadableArray())
                    putArray("routeProfits", outRouteProfits.take(nRoutes).toReadableArray())
                    putArray("margins", outMargins.take(nRoutes).toReadableArray())
                    putInt("nConstraints", nConstraints)
                    putArray("shadowPrices", outShadowPrices.take(nConstraints).toReadableArray())
                }

                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("SOLVER_ERROR", e.message, e)
            }
        }
    }

    // ── monteCarlo ─────────────────────────────────────────────────────────────

    @ReactMethod
    fun monteCarlo(
        descriptorBase64: String,
        centerArray: ReadableArray,
        nScenarios: Int,
        promise: Promise
    ) {
        scope.launch {
            try {
                val descriptor = Base64.decode(descriptorBase64, Base64.DEFAULT)
                val center = centerArray.toDoubleArray()

                val outStatus = IntArray(1)
                val outCounts = IntArray(3)
                val outStats = DoubleArray(9)
                val outNSens = IntArray(1)
                val outSensIdx = IntArray(64)
                val outSensCorr = DoubleArray(64)

                nativeMonteCarlo(
                    descriptor, descriptor.size,
                    center, center.size,
                    nScenarios,
                    outStatus, outCounts, outStats,
                    outNSens, outSensIdx, outSensCorr
                )

                val nSens = outNSens[0]

                val result = Arguments.createMap().apply {
                    putInt("status", outStatus[0])
                    putInt("nScenarios", outCounts[0])
                    putInt("nFeasible", outCounts[1])
                    putInt("nInfeasible", outCounts[2])
                    putDouble("mean", outStats[0])
                    putDouble("stddev", outStats[1])
                    putDouble("p5", outStats[2])
                    putDouble("p25", outStats[3])
                    putDouble("p50", outStats[4])
                    putDouble("p75", outStats[5])
                    putDouble("p95", outStats[6])
                    putDouble("min", outStats[7])
                    putDouble("max", outStats[8])
                    putArray("sensitivityIdx", outSensIdx.take(nSens).toReadableArray())
                    putArray("sensitivityCorr", outSensCorr.take(nSens).toReadableArray())
                }

                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("MONTE_CARLO_ERROR", e.message, e)
            }
        }
    }

    // ── getVersion ─────────────────────────────────────────────────────────────

    @ReactMethod
    fun getVersion(promise: Promise) {
        promise.resolve(nativeGetVersion())
    }

    // ── Helpers ────────────────────────────────────────────────────────────────

    private fun ReadableArray.toDoubleArray(): DoubleArray =
        DoubleArray(size()) { i -> getDouble(i) }

    private fun List<Double>.toReadableArray(): WritableArray =
        Arguments.createArray().also { arr -> forEach { arr.pushDouble(it) } }

    private fun List<Int>.toReadableArray(): WritableArray =
        Arguments.createArray().also { arr -> forEach { arr.pushInt(it) } }

    private fun DoubleArray.take(n: Int) = toList().subList(0, n.coerceAtMost(size))
    private fun IntArray.take(n: Int) = toList().subList(0, n.coerceAtMost(size))
}

File: mobile/TraderDesk/ios/TradingSolverModule.swift
// TradingSolverModule.swift
// React Native bridge to the Zig LP solver static library (libtrading_solver.a)
//
// The Zig library exposes a C ABI (see native/solver_mobile.zig).
// Swift calls it via @_silgen_name / UnsafePointer bridging through
// the solver_mobile.h C header.

import Foundation

// ── C struct mirrors (must match solver_mobile.zig extern structs) ────────────

struct CMobileSolveResult {
  var status:        Int32
  var profit:        Double
  var tons:          Double
  var cost:          Double
  var roi:           Double
  var n_routes:      Int32
  var route_tons:    (Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double)
  var route_profits: (Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double)
  var margins:       (Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double)
  var n_constraints: Int32
  var shadow_prices: (Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double,
                       Double, Double, Double, Double, Double, Double, Double, Double)
}

struct CMobileMonteCarloResult {
  var status:           Int32
  var n_scenarios:      UInt32
  var n_feasible:       UInt32
  var n_infeasible:     UInt32
  var mean:             Double
  var stddev:           Double
  var p5:               Double
  var p25:              Double
  var p50:              Double
  var p75:              Double
  var p95:              Double
  var min:              Double
  var max:              Double
  var n_sensitivity:    Int32
  var sensitivity_idx:  (Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
                          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32)
  var sensitivity_corr: (Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double,
                          Double, Double, Double, Double, Double, Double, Double, Double)
}

// ── C function declarations ───────────────────────────────────────────────────
// These link to the exported symbols from libtrading_solver.a

@_silgen_name("trading_solve")
func trading_solve(
  _ model_descriptor: UnsafePointer<UInt8>,
  _ model_len: Int,
  _ variables: UnsafePointer<Double>,
  _ n_vars: Int,
  _ out: UnsafeMutablePointer<CMobileSolveResult>
) -> Int32

@_silgen_name("trading_monte_carlo")
func trading_monte_carlo(
  _ model_descriptor: UnsafePointer<UInt8>,
  _ model_len: Int,
  _ center: UnsafePointer<Double>,
  _ n_vars: Int,
  _ n_scenarios: UInt32,
  _ out: UnsafeMutablePointer<CMobileMonteCarloResult>
) -> Int32

@_silgen_name("trading_solver_version")
func trading_solver_version() -> UnsafePointer<CChar>

// ── React Native module ───────────────────────────────────────────────────────

@objc(TradingSolver)
class TradingSolverModule: NSObject {

  @objc static func requiresMainQueueSetup() -> Bool { false }

  // ── solve ──────────────────────────────────────────────────────────────────

  @objc func solve(
    _ descriptorBase64: String,
    variables variablesArray: [Double],
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      guard let descriptorData = Data(base64Encoded: descriptorBase64) else {
        reject("INVALID_DESCRIPTOR", "Could not base64-decode model descriptor", nil)
        return
      }

      var result = CMobileSolveResult(
        status: 0, profit: 0, tons: 0, cost: 0, roi: 0,
        n_routes: 0,
        route_tons: (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0),
        route_profits: (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0),
        margins: (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0),
        n_constraints: 0,
        shadow_prices: (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                         0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
      )

      let status = descriptorData.withUnsafeBytes { descPtr in
        trading_solve(
          descPtr.bindMemory(to: UInt8.self).baseAddress!,
          descriptorData.count,
          variablesArray,
          variablesArray.count,
          &result
        )
      }

      let nRoutes = Int(result.n_routes)
      let nConstraints = Int(result.n_constraints)

      let routeTons = Self.tupleToArray16(result.route_tons, count: nRoutes)
      let routeProfits = Self.tupleToArray16(result.route_profits, count: nRoutes)
      let margins = Self.tupleToArray16(result.margins, count: nRoutes)
      let shadowPrices = Self.tupleToArray32(result.shadow_prices, count: nConstraints)

      resolve([
        "status": Int(status),
        "profit": result.profit,
        "tons": result.tons,
        "cost": result.cost,
        "roi": result.roi,
        "nRoutes": nRoutes,
        "routeTons": routeTons,
        "routeProfits": routeProfits,
        "margins": margins,
        "nConstraints": nConstraints,
        "shadowPrices": shadowPrices,
      ])
    }
  }

  // ── monteCarlo ─────────────────────────────────────────────────────────────

  @objc func monteCarlo(
    _ descriptorBase64: String,
    centerVariables: [Double],
    nScenarios: Int,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    DispatchQueue.global(qos: .userInitiated).async {
      guard let descriptorData = Data(base64Encoded: descriptorBase64) else {
        reject("INVALID_DESCRIPTOR", "Could not base64-decode model descriptor", nil)
        return
      }

      var result = CMobileMonteCarloResult(
        status: 0, n_scenarios: 0, n_feasible: 0, n_infeasible: 0,
        mean: 0, stddev: 0, p5: 0, p25: 0, p50: 0, p75: 0, p95: 0,
        min: 0, max: 0, n_sensitivity: 0,
        sensitivity_idx:  (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0),
        sensitivity_corr: (0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                            0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
      )

      let status = descriptorData.withUnsafeBytes { descPtr in
        trading_monte_carlo(
          descPtr.bindMemory(to: UInt8.self).baseAddress!,
          descriptorData.count,
          centerVariables,
          centerVariables.count,
          UInt32(nScenarios),
          &result
        )
      }

      let nSens = Int(result.n_sensitivity)
      let sensIdx = Self.tupleToArray64Int(result.sensitivity_idx, count: nSens)
      let sensCorr = Self.tupleToArray64(result.sensitivity_corr, count: nSens)

      resolve([
        "status": Int(status),
        "nScenarios": Int(result.n_scenarios),
        "nFeasible": Int(result.n_feasible),
        "nInfeasible": Int(result.n_infeasible),
        "mean": result.mean,
        "stddev": result.stddev,
        "p5": result.p5,
        "p25": result.p25,
        "p50": result.p50,
        "p75": result.p75,
        "p95": result.p95,
        "min": result.min,
        "max": result.max,
        "sensitivityIdx": sensIdx,
        "sensitivityCorr": sensCorr,
      ])
    }
  }

  // ── getVersion ─────────────────────────────────────────────────────────────

  @objc func getVersion(
    _ resolve: @escaping RCTPromiseResolveBlock,
    reject _: @escaping RCTPromiseRejectBlock
  ) {
    let ptr = trading_solver_version()
    resolve(String(cString: ptr))
  }

  // ── Array helpers (Swift tuple → [Double/Int32]) ──────────────────────────

  private static func tupleToArray16(
    _ t: (Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double),
    count: Int
  ) -> [Double] {
    let arr = [t.0,t.1,t.2,t.3,t.4,t.5,t.6,t.7,t.8,t.9,t.10,t.11,t.12,t.13,t.14,t.15]
    return Array(arr.prefix(count))
  }

  private static func tupleToArray32(
    _ t: (Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double),
    count: Int
  ) -> [Double] {
    let arr = [t.0,t.1,t.2,t.3,t.4,t.5,t.6,t.7,t.8,t.9,t.10,t.11,t.12,t.13,t.14,t.15,
               t.16,t.17,t.18,t.19,t.20,t.21,t.22,t.23,t.24,t.25,t.26,t.27,t.28,t.29,t.30,t.31]
    return Array(arr.prefix(count))
  }

  private static func tupleToArray64(
    _ t: (Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double,
          Double, Double, Double, Double, Double, Double, Double, Double),
    count: Int
  ) -> [Double] {
    var arr: [Double] = []
    let mirror = Mirror(reflecting: t)
    for child in mirror.children { arr.append(child.value as! Double) } // swiftlint:disable:this force_cast
    return Array(arr.prefix(count))
  }

  private static func tupleToArray64Int(
    _ t: (Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32,
          Int32, Int32, Int32, Int32, Int32, Int32, Int32, Int32),
    count: Int
  ) -> [Int32] {
    var arr: [Int32] = []
    let mirror = Mirror(reflecting: t)
    for child in mirror.children { arr.append(child.value as! Int32) } // swiftlint:disable:this force_cast
    return Array(arr.prefix(count))
  }
}
